File: douyin-operations-shortvideo/src/main/java/cn/gaifan/douyinOperations/module/shortvideo/entity/SvAccount.java
package cn.gaifan.douyinOperations.module.shortvideo.entity;

import lombok.Data;
import org.hibernate.annotations.SQLRestriction;

import jakarta.persistence.*;
import java.math.BigDecimal;
import java.sql.Timestamp;

/**
 * 短视频账号主表
 * <p>
 * 统一管理所有采集的账号信息，支持账号级别的数据分析和统计。
 * 一个账号（sec_uid）对应多个采集任务和多个爆款视频。
 */
@Data
@Entity
@Table(name = "sv_account")
@SQLRestriction("deleted = 0")
public class SvAccount {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "owner_id", nullable = false)
    private Long ownerId;

    // ─── 账号基本信息 ──────────────────────────────────────

    /** 抖音唯一标识（全局唯一） */
    @Column(name = "sec_uid", nullable = false, length = 256)
    private String secUid;

    /** 抖音号 */
    @Column(name = "douyin_id", length = 128)
    private String douyinId;

    /** 昵称 */
    @Column(name = "nickname", length = 128)
    private String nickname;

    /** 头像 URL */
    @Column(name = "avatar_url", length = 512)
    private String avatarUrl;

    /** 个人简介 */
    @Column(name = "signature", columnDefinition = "TEXT")
    private String signature;

    // ─── 账号数据 ──────────────────────────────────────

    /** 粉丝数 */
    @Column(name = "follower_count")
    private Long followerCount = 0L;

    /** 关注数 */
    @Column(name = "following_count")
    private Long followingCount = 0L;

    /** 获赞总数 */
    @Column(name = "total_favorited")
    private Long totalFavorited = 0L;

    /** 作品数 */
    @Column(name = "video_count")
    private Integer videoCount = 0;

    // ─── 认证信息 ──────────────────────────────────────

    /** 是否认证 */
    @Column(name = "is_verified")
    private Boolean isVerified = false;

    /** 认证类型：personal/enterprise/government */
    @Column(name = "verification_type", length = 32)
    private String verificationType;

    // ─── 采集统计 ──────────────────────────────────────

    /** 采集次数 */
    @Column(name = "collect_count")
    private Integer collectCount = 0;

    /** 最后采集时间 */
    @Column(name = "last_collect_time")
    private Timestamp lastCollectTime;

    /** 累计采集视频数 */
    @Column(name = "total_collected_videos")
    private Integer totalCollectedVideos = 0;

    // ─── 分析统计 ──────────────────────────────────────

    /** 平均播放量 */
    @Column(name = "avg_view_count")
    private Long avgViewCount = 0L;

    /** 平均点赞数 */
    @Column(name = "avg_like_count")
    private Integer avgLikeCount = 0;

    /** 平均分享数 */
    @Column(name = "avg_share_count")
    private Integer avgShareCount = 0;

    /** 平均评论数 */
    @Column(name = "avg_comment_count")
    private Integer avgCommentCount = 0;

    /** 平均爆款评分 */
    @Column(name = "avg_viral_score", precision = 5, scale = 2)
    private BigDecimal avgViralScore = BigDecimal.ZERO;

    /** 最高爆款评分 */
    @Column(name = "top_viral_score", precision = 5, scale = 2)
    private BigDecimal topViralScore = BigDecimal.ZERO;

    // ─── 标签与分类 ──────────────────────────────────────

    /** 行业标签（JSON 数组） */
    @Column(name = "industry_tags", length = 512)
    private String industryTags;

    /** 内容标签（JSON 数组） */
    @Column(name = "content_tags", length = 512)
    private String contentTags;

    /** 账号分类 */
    @Column(name = "account_category", length = 64)
    private String accountCategory;

    // ─── 来源信息 ──────────────────────────────────────

    /** 来源类型：manual/keyword_search/recommend/import */
    @Column(name = "source_type", length = 32)
    private String sourceType = "manual";

    /** 来源关键词（关键词采集时记录） */
    @Column(name = "source_keyword", length = 256)
    private String sourceKeyword;

    /** 来源采集任务 ID */
    @Column(name = "source_task_id")
    private Long sourceTaskId;

    // ─── 备注与状态 ──────────────────────────────────────

    /** 备注 */
    @Column(name = "notes", columnDefinition = "TEXT")
    private String notes;

    /** 账号状态：active/archived/blocked */
    @Column(name = "status", length = 32)
    private String status = "active";

    // ─── 系统字段 ──────────────────────────────────────

    @Column(name = "deleted", nullable = false)
    private Integer deleted = 0;

    @Column(name = "create_time")
    private Timestamp createTime;

    @Column(name = "update_time")
    private Timestamp updateTime;

    @PrePersist
    public void prePersist() {
        if (createTime == null) createTime = new Timestamp(System.currentTimeMillis());
        if (updateTime == null) updateTime = new Timestamp(System.currentTimeMillis());
    }

    @PreUpdate
    public void preUpdate() {
        updateTime = new Timestamp(System.currentTimeMillis());
    }
}

File: front/src/api/sv-account.ts
import request from '@/utils/request'
import type { PageResult } from '@/types/common'
import {
  isRecord,
  normalizeRecord as normalizeResponseRecord,
  normalizePage as normalizeResponsePage,
  readNumber,
} from '@/utils/response-normalize'

// ─── 类型定义 ──────────────────────────────────────

export interface SvAccount {
  id: number
  secUid: string
  douyinId?: string
  nickname?: string
  avatarUrl?: string
  signature?: string

  // 账号数据
  followerCount: number
  followingCount: number
  totalFavorited: number
  videoCount: number

  // 认证信息
  isVerified: boolean
  verificationType?: string

  // 采集统计
  collectCount: number
  lastCollectTime?: string
  totalCollectedVideos: number

  // 分析统计
  avgViewCount: number
  avgLikeCount: number
  avgViralScore: number
  topViralScore: number

  // 标签与分类
  industryTags?: string
  contentTags?: string
  accountCategory?: string

  // 来源信息
  sourceType: string
  sourceKeyword?: string

  // 状态
  status: string
  notes?: string

  createTime: string
  updateTime: string
}

export interface SvAccountDetail extends SvAccount {
  taskCount: number
  pendingAnalysisCount: number
  analyzedCount: number
  latestTaskId?: number
}

export interface SvAccountSearchParams {
  keyword?: string
  accountCategory?: string
  sourceType?: string
  sourceKeyword?: string
  status?: string
  minFollowerCount?: number
  minViralScore?: number
  page: number
  rows: number
  sortName?: string
  sortOrder?: string
}

export interface SvAccountUpdateParams {
  id: number
  accountCategory?: string
  industryTags?: string
  contentTags?: string
  notes?: string
  status?: string
}

export interface AccountVideo {
  id: number
  title: string
  coverUrl?: string
  videoUrl?: string
  douyinVideoId?: string
  authorName?: string
  viewCount: number
  likeCount: number
  shareCount: number
  viralScore: number
  deepAnalyzeStatus: string
  /** 列表接口仍可能返回短进度摘要 */
  deepAnalyzeProgress?: string | null
  publishTime?: string
  createTime: string
  updateTime?: string
}

/** 与后端 AccountVideosQueryVO 一致 */
export interface AccountVideosQueryParams {
  accountId: number
  page?: number
  rows?: number
  keyword?: string
  deepAnalyzeStatus?: string
  sortName?: 'createTime' | 'viewCount' | 'viralScore' | 'updateTime' | 'id'
  sortOrder?: 'asc' | 'desc'
}

/** 与后端 SvAccountAnalyticsVO 对齐：该账号下已采集视频聚合 */
export interface SvAccountAnalytics {
  videoCount: number
  sumViewCount: number
  sumLikeCount: number
  sumShareCount: number
  avgViewPerVideo: number
  avgLikePerVideo: number
  avgSharePerVideo: number
  avgViralScore: number
  maxViewCount: number
  minViewCount: number
  deepPendingCount: number
  deepProcessingCount: number
  deepCompletedCount: number
  deepFailedCount: number
  deepOtherCount: number
}

function readOptionalNumber(value: unknown): number | undefined {
  if (value == null || value === '') return undefined
  const n = Number(value)
  return Number.isFinite(n) ? n : undefined
}

function readStringAlias(row: Record<string, unknown>, keys: string[], fallback = ''): string {
  for (const key of keys) {
    const value = row[key]
    if (typeof value === 'string' && value.trim()) return value.trim()
    if (typeof value === 'number' && Number.isFinite(value)) return String(value)
  }
  return fallback
}

function readOptionalStringAlias(row: Record<string, unknown>, keys: string[]): string | undefined {
  const value = readStringAlias(row, keys)
  return value || undefined
}

function readBooleanAlias(value: unknown, fallback = false): boolean {
  if (typeof value === 'boolean') return value
  if (typeof value === 'number') return value !== 0
  if (typeof value === 'string') {
    const text = value.trim().toLowerCase()
    if (['true', '1', 'yes', 'y', 'verified'].includes(text)) return true
    if (['false', '0', 'no', 'n', 'unverified'].includes(text)) return false
  }
  return fallback
}

function unwrapRecord(raw: unknown): Record<string, unknown> {
  const value = normalizeResponseRecord(raw)
  return isRecord(value) ? value : {}
}

function normalizeAccount(raw: unknown): SvAccount {
  const row = unwrapRecord(raw)
  return {
    ...(row as unknown as SvAccount),
    id: readNumber(row, ['id', 'accountId'], 0),
    secUid: readStringAlias(row, ['secUid', 'sec_uid', 'secUserId']),
    douyinId: readOptionalStringAlias(row, ['douyinId', 'douyin_id', 'uniqueId']),
    nickname: readOptionalStringAlias(row, ['nickname', 'nickName', 'authorName']),
    avatarUrl: readOptionalStringAlias(row, ['avatarUrl', 'avatar', 'authorAvatar']),
    signature: readOptionalStringAlias(row, ['signature', 'description', 'bio']),
    followerCount: readNumber(row, ['followerCount', 'followers', 'fansCount'], 0),
    followingCount: readNumber(row, ['followingCount', 'followings', 'following'], 0),
    totalFavorited: readNumber(row, ['totalFavorited', 'favoritedCount', 'totalLikeCount'], 0),
    videoCount: readNumber(row, ['videoCount', 'awemeCount', 'worksCount'], 0),
    isVerified: readBooleanAlias(row.isVerified ?? row.verified),
    verificationType: readOptionalStringAlias(row, ['verificationType', 'verifyType']),
    collectCount: readNumber(row, ['collectCount', 'collectionCount'], 0),
    lastCollectTime: readOptionalStringAlias(row, ['lastCollectTime', 'lastCollectedAt']),
    totalCollectedVideos: readNumber(row, ['totalCollectedVideos', 'collectedVideoCount', 'viralVideoCount'], 0),
    avgViewCount: readNumber(row, ['avgViewCount', 'averageViewCount'], 0),
    avgLikeCount: readNumber(row, ['avgLikeCount', 'averageLikeCount'], 0),
    avgViralScore: readNumber(row, ['avgViralScore', 'averageViralScore'], 0),
    topViralScore: readNumber(row, ['topViralScore', 'maxViralScore'], 0),
    industryTags: readOptionalStringAlias(row, ['industryTags', 'industryTag']),
    contentTags: readOptionalStringAlias(row, ['contentTags', 'contentTag']),
    accountCategory: readOptionalStringAlias(row, ['accountCategory', 'category']),
    sourceType: readStringAlias(row, ['sourceType', 'source'], 'unknown'),
    sourceKeyword: readOptionalStringAlias(row, ['sourceKeyword', 'keyword']),
    status: readStringAlias(row, ['status'], 'active'),
    notes: readOptionalStringAlias(row, ['notes', 'remark']),
    createTime: readStringAlias(row, ['createTime', 'createdAt']),
    updateTime: readStringAlias(row, ['updateTime', 'updatedAt']),
  }
}

function normalizeAccountDetail(raw: unknown): SvAccountDetail {
  const row = unwrapRecord(raw)
  return {
    ...normalizeAccount(row),
    taskCount: readNumber(row, ['taskCount', 'collectTaskCount'], 0),
    pendingAnalysisCount: readNumber(row, ['pendingAnalysisCount', 'deepPendingCount'], 0),
    analyzedCount: readNumber(row, ['analyzedCount', 'deepCompletedCount'], 0),
    latestTaskId: readOptionalNumber(row.latestTaskId ?? row.lastTaskId),
  }
}

function normalizeAccountVideo(raw: unknown): AccountVideo {
  const row = unwrapRecord(raw)
  return {
    ...(row as unknown as AccountVideo),
    id: readNumber(row, ['id', 'videoId', 'viralVideoId'], 0),
    title: readStringAlias(row, ['title', 'desc', 'description']),
    coverUrl: readOptionalStringAlias(row, ['coverUrl', 'cover', 'coverImage']),
    videoUrl: readOptionalStringAlias(row, ['videoUrl', 'url', 'playUrl']),
    douyinVideoId: readOptionalStringAlias(row, ['douyinVideoId', 'awemeId']),
    authorName: readOptionalStringAlias(row, ['authorName', 'nickname']),
    viewCount: readNumber(row, ['viewCount', 'playCount'], 0),
    likeCount: readNumber(row, ['likeCount', 'diggCount'], 0),
    shareCount: readNumber(row, ['shareCount'], 0),
    viralScore: readNumber(row, ['viralScore', 'score'], 0),
    deepAnalyzeStatus: readStringAlias(row, ['deepAnalyzeStatus', 'analysisStatus'], 'pending'),
    deepAnalyzeProgress: readOptionalStringAlias(row, ['deepAnalyzeProgress', 'analysisProgress']),
    publishTime: readOptionalStringAlias(row, ['publishTime', 'publishedAt']),
    createTime: readStringAlias(row, ['createTime', 'createdAt']),
    updateTime: readOptionalStringAlias(row, ['updateTime', 'updatedAt']),
  }
}

function normalizeAnalytics(raw: unknown): SvAccountAnalytics {
  const row = unwrapRecord(raw)
  return {
    videoCount: readNumber(row, ['videoCount', 'totalVideoCount'], 0),
    sumViewCount: readNumber(row, ['sumViewCount', 'totalViewCount'], 0),
    sumLikeCount: readNumber(row, ['sumLikeCount', 'totalLikeCount'], 0),
    sumShareCount: readNumber(row, ['sumShareCount', 'totalShareCount'], 0),
    avgViewPerVideo: readNumber(row, ['avgViewPerVideo', 'avgViewCount'], 0),
    avgLikePerVideo: readNumber(row, ['avgLikePerVideo', 'avgLikeCount'], 0),
    avgSharePerVideo: readNumber(row, ['avgSharePerVideo', 'avgShareCount'], 0),
    avgViralScore: readNumber(row, ['avgViralScore', 'averageViralScore'], 0),
    maxViewCount: readNumber(row, ['maxViewCount', 'topViewCount'], 0),
    minViewCount: readNumber(row, ['minViewCount'], 0),
    deepPendingCount: readNumber(row, ['deepPendingCount', 'pendingAnalysisCount'], 0),
    deepProcessingCount: readNumber(row, ['deepProcessingCount', 'processingAnalysisCount'], 0),
    deepCompletedCount: readNumber(row, ['deepCompletedCount', 'analyzedCount'], 0),
    deepFailedCount: readNumber(row, ['deepFailedCount', 'failedAnalysisCount'], 0),
    deepOtherCount: readNumber(row, ['deepOtherCount', 'otherAnalysisCount'], 0),
  }
}

// ─── API 方法 ──────────────────────────────────────

/**
 * 账号列表
 */
export function accountList(params: SvAccountSearchParams) {
  return request
    .post<unknown>('/short-video/account/list', params)
    .then((raw): PageResult<SvAccount> => normalizeResponsePage<unknown, SvAccount>(
      raw,
      normalizeAccount,
      params.page,
      params.rows,
    ))
}

/**
 * 账号详情
 */
export function accountGet(id: number) {
  return request.post<unknown>('/short-video/account/get', { id }).then(normalizeAccountDetail)
}

/**
 * 更新账号
 */
export function accountUpdate(params: SvAccountUpdateParams) {
  return request.post<void>('/short-video/account/update', params)
}

/**
 * 删除账号
 */
export function accountDelete(id: number) {
  return request.post<void>('/short-video/account/delete', { id })
}

/**
 * 账号视频列表（分页、筛选、排序）
 */
export function accountVideos(params: AccountVideosQueryParams) {
  return request
    .post<unknown>('/short-video/account/videos', params)
    .then((raw): PageResult<AccountVideo> => normalizeResponsePage<unknown, AccountVideo>(
      raw,
      normalizeAccountVideo,
      params.page ?? 0,
      params.rows ?? 20,
    ))
}

/**
 * 账号下全部已采集视频的聚合分析（播放、点赞、深度分析状态分布等）
 */
export function accountAnalytics(accountId: number) {
  return request.post<unknown>('/short-video/account/analytics', { id: accountId }).then(normalizeAnalytics)
}

/**
 * 刷新账号统计
 */
export function accountRefreshStats(id: number) {
  return request.post<void>('/short-video/account/refresh-stats', { id })
}
